url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/gui/GUIScreen.java

```diff
@@ -0,0 +1,410 @@
+                preferredSize.setRows(maxSizeHeight);
+            
+            if(windowPlacement.getPositionPolicy() == Position.FULL_SCREEN) {
+                preferredSize.setColumns(preferredSize.getColumns() + 1);
+                preferredSize.setRows(preferredSize.getRows() + 1);
+            }
+            
+            if(topLeft.getColumn() < 0)
+                topLeft.setColumn(0);
+            if(topLeft.getRow() < 0)
+                topLeft.setRow(0);
+
+            TextGraphics subGraphics = textGraphics.subAreaGraphics(topLeft,
+                    new TerminalSize(preferredSize.getColumns(), preferredSize.getRows()));
+
+            //First draw the shadow
+            textGraphics.applyTheme(guiTheme.getDefinition(Theme.Category.Shadow));
+            textGraphics.fillRectangle(' ', new TerminalPosition(topLeft.getColumn() + 2, topLeft.getRow() + 1),
+                    new TerminalSize(subGraphics.getWidth(), subGraphics.getHeight()));
+
+            //Then draw the window
+            windowPlacement.getWindow().repaint(subGraphics);
+        }
+
+        if(windowStack.size() > 0 && windowStack.getLast().getWindow().getWindowHotspotPosition() != null)
+            screen.setCursorPosition(windowStack.getLast().getWindow().getWindowHotspotPosition());
+        else
+            screen.setCursorPosition(new TerminalPosition(screenSizeColumns - 1, screenSizeRows - 1));
+        screen.refresh();
+    }
+
+    private boolean update()
+    {
+        if(needsRefresh || screen.resizePending()) {
+            repaint();
+            needsRefresh = false;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Signals the the entire screen needs to be re-drawn
+     */
+    public void invalidate()
+    {
+        needsRefresh = true;
+    }
+
+    private void doEventLoop()
+    {
+        int currentStackLength = windowStack.size();
+        if(currentStackLength == 0)
+            return;
+
+        while(true) {
+            if(currentStackLength > windowStack.size()) {
+                //The window was removed from the stack ( = it was closed)
+                break;
+            }
+
+            synchronized(actionToRunInEventThread) {
+                List<Action> actions = new ArrayList<Action>(actionToRunInEventThread);
+                actionToRunInEventThread.clear();
+                for(Action nextAction: actions)
+                    nextAction.doAction();
+            }
+
+            boolean repainted = update();
+
+            Key nextKey = screen.readInput();
+            if(nextKey != null) {
+                windowStack.getLast().window.onKeyPressed(nextKey);
+                invalidate();
+            }
+            else {
+                if(!repainted) {
+                    try {
+                        Thread.sleep(1);
+                    }
+                    catch(InterruptedException e) {}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)